File: octave/genesis/least-squares/least-squares.m
# Solving Bevington example 6.1 with Julia
printf( "Bevington Example 6.1\n" )
strftime ("%Y-%m-%d %H:%M:%S", localtime (time ()))

# design matrix
A = [ 1 1; 1 2; 1 3; 1 4; 1 5; 1 6; 1 7; 1 8; 1 9 ];
printf( "design matrix:\n" )
A
# disp(A)

# Define data
x = [ 1; 2; 3; 4; 5; 6; 7; 8; 9 ];
#T = [ 78/5; 35/2; 183/5; 219/5; 291/5; 308/5; 321/5; 352/5; 494/5 ];
T = [15.6; 17.5; 36.6; 43.8; 58.2; 61.6; 64.2; 70.4; 98.8];
printf( "data vector:\n" )
T

# solve least squares problem
xls = A \ T;
printf( "least squares solution\n" )
printf( "xls = A \\ T:\n" )
xls

# error analysis

# residual error vector
residual = A * xls - T;
printf( "residual error vector\n" )
printf( "residual = A * xls - T:\n" )

t2 = dot( residual, residual );
printf( "least total squared error\n" )
printf( "t2 = residual . residual :\n" )
t2

# compute Gram matrix
W = transpose( A ) * A
printf( "compute Gram matrix\n" )
printf( "W = transpose( A ) * A\n" )

Winv = inv( W )
printf( "invert Gram matrix\n" )
printf( "Winv = inv( W )\n" )
Winv

values = diag( Winv )
printf( "harvest diagonal elements of Winv" )
printf( "values = diag( Winv ):\n" )
values

# measure design matrix
# m = rows, n = columns
[ m, n ] = size ( A );
g = sprintf('%d ', [ m, n ]);
fprintf('matrix dimensions: %s\n', g);

sigma = sqrt( t2 / ( m - n ) * values );
printf( "compute error elements:\n" )
printf( "sigma = sqrt( t2 / ( m - n ) * values ):\n" )
sigma

printf( "\n#  #  #  Compare Octave values to exact values\n" )
printf( "\nFit parameters\n" )

numericError = xls - [ 1733 / 360; 1129 / 120 ];
printf( "\nerror in intercept and slope values\n" )
printf( "numericError = xls - [ 1733 / 360; 1129 / 120 ]\n" )
numericError

epsError = numericError ./ eps( 1.0 );
printf( "\nerror in intercept and slope values in machine epsilon\n" )
printf( "epsError = numericError ./ eps( 1.0 )\n" )
epsError

printf( "\nError parameters\n" )
numericError = sigma - sqrt( [ 108297055; 3419907 ] / 35 ) / 360;
printf( "\nintercept and slope sigmas\n" )
printf( "numericError = sigma - sqrt( [ 108297055; 3419907 ] / 35 ) / 360\n" )
numericError

epsError = numericError ./ eps( 1.0 );
printf( "\nerror in intercept and slope sigmas in machine epsilon\n" )
printf( "epsError = numericError ./ eps( 1.0 )\n" )
epsError

output_precision(16)
printf( "\noutput_precision(16)\n" )
xls
sigma

// /Volumes/T7-Touch/repos/github/jop/octave/genesis/least-squares
// dantopa@Quaxolotl.local:least-squares $ octave-cli least-squares.m
// Bevington Example 6.1
// ans = 2022-09-01 05:22:36
// design matrix:
// A =
//
//    1   1
//    1   2
//    1   3
//    1   4
//    1   5
//    1   6
//    1   7
//    1   8
//    1   9
//
// data vector:
// T =
//
//    15.600
//    17.500
//    36.600
//    43.800
//    58.200
//    61.600
//    64.200
//    70.400
//    98.800
//
// least squares solution
// xls = A \ T:
// xls =
//
//    4.8139
//    9.4083
//
// residual error vector
// residual = A * xls - T:
// least total squared error
// t2 = residual . residual :
// t2 = 316.66
// W =
//
//      9    45
//     45   285
//
// compute Gram matrix
// W = transpose( A ) * A
// Winv =
//
//    0.527778  -0.083333
//   -0.083333   0.016667
//
// invert Gram matrix
// Winv = inv( W )
// Winv =
//
//    0.527778  -0.083333
//   -0.083333   0.016667
//
// values =
//
//    0.527778
//    0.016667
//
// harvest diagonal elements of Winvvalues = diag( Winv ):
// values =
//
//    0.527778
//    0.016667
//
// matrix dimensions: 9 2
// compute error elements:
// sigma = sqrt( t2 / ( m - n ) * values ):
// sigma =
//
//    4.8862
//    0.8683
//
//
// #  #  #  Compare Octave values to exact values
//
// Fit parameters
//
// error in intercept and slope values
// numericError = xls - [ 1733 / 360; 1129 / 120 ]
// numericError =
//
//   -1.0658e-14
//    1.7764e-15
//
//
// error in intercept and slope values in machine epsilon
// epsError = numericError ./ eps( 1.0 )
// epsError =
//
//   -48
//     8
//
//
// Error parameters
//
// intercept and slope sigmas
// numericError = sigma - sqrt( [ 108297055; 3419907 ] / 35 ) / 360
// numericError =
//
//    8.8818e-16
//    1.1102e-16
//
//
// error in intercept and slope sigmas in machine epsilon
// epsError = numericError ./ eps( 1.0 )
// epsError =
//
//    4.0000
//    0.5000
//
//
// output_precision(16)
// xls =
//
//    4.813888888888878
//    9.408333333333335
//
// sigma =
//
//    4.886206312183355
//    0.868301647656361
//

File: python/genesis/least-squares/least-squares.py
#! /opt/local/bin/python
#! /home/dantopa/spacktivity/ubuntu-22.04-dantopa-docker-spack/opt/spack/linux-ubuntu22.04-haswell/gcc-12.2.0/python-3.10.6-ybcookynohghdr6i72gwalqle6hq27z5/bin/python
# Solving Bevington example 6.1 with Julia
from datetime import date
today = date.today( )
print( today.strftime("%b-%d-%Y") )
import os
print(os.path.dirname(os.path.realpath(__file__)))

print( "importing numpy..." )
import numpy as np

# input data
A = np.array([ [1, 1], [1, 2], [1, 3], [1, 4], [1, 5], [1, 6], [1, 7], [1, 8], [1, 9] ])
print( "design matrix A\n", np.matrix( A ) )
T = np.array([15.6, 17.5, 36.6, 43.8, 58.2, 61.6, 64.2, 70.4, 98.8])
print ( "data vector T =", np.matrix( T ))

# set up normal equations
W = np.dot( A.transpose(), A )
print( "Gram matrix A*A\n", np.matrix( W ) )
b = np.dot( A.transpose(), T )

# solve normal equations
xls = np.linalg.solve( W, b )
print ( "least squares solution xls =", xls )

# pseudoinverse solution
Winv  = np.linalg.inv( W )
Ap    = np.dot( Winv, A.transpose( ) )
xpinv = np.dot( Ap, T.transpose( ) )
print( "pseudoinverse        xpinv =", xpinv )

# error propagation
residual = np.dot( A, xls ) - T
t2 = np.dot( residual, residual )
print( "least total squared error =", t2 )
[ m, n ] = A.shape
print( "A matrix dimensions =", [ m, n ] )
diag  = np.diag( Winv )
sigma = np.sqrt( t2 / ( m - n ) * diag )
print ( "sigma =", sigma )

# numerical errors
print( "\nnumerical errors\n" )
xlsExact = [ 1733 / 360, 1129 / 120 ]
numericErrorXLS = xls - xlsExact
print( "numeric errors, fit parameters" )
print( numericErrorXLS )

sigmaExact = np.sqrt( np.array( [ 108297055, 3419907 ] ) / 35 ) / 360
numericErrorSigma = sigma - sigmaExact
print( "\nnumeric errors, sigma parameters" )
print( numericErrorSigma )


# machine epsilon
machineEpsilon = np.finfo(float).eps
print( "\nmachine epsilon =", np.finfo(float).eps )
numericErrorXLSeps   = numericErrorXLS   / machineEpsilon
numericErrorXLSsigma = numericErrorSigma / machineEpsilon
print( "\nnumeric epsilons, fit parameters:   ", numericErrorXLSeps )
print( "\nnumeric epsilons, sigma parameters: ", numericErrorXLSsigma )

# dantopa@Quaxolotl.attlocal.net:least-squares $ ./least-squares.py
# Sep-05-2022
# /Volumes/T7-Touch/repos/github/jop/python/genesis/least-squares

# importing numpy...
# design matrix A
#  [[1 1]
#  [1 2]
#  [1 3]
#  [1 4]
#  [1 5]
#  [1 6]
#  [1 7]
#  [1 8]
#  [1 9]]
# data vector T = [[15.6 17.5 36.6 43.8 58.2 61.6 64.2 70.4 98.8]]
# Gram matrix A*A
#  [[  9  45]
#  [ 45 285]]
# least squares solution xls = [4.81388889 9.40833333]
# pseudoinverse        xpinv = [4.81388889 9.40833333]
# least total squared error = 316.6580555555554
# A matrix dimensions = [9, 2]
# sigma = [4.88620631 0.86830165]
#
# numerical errors
#
# numeric errors, fit parameters
# [ 9.76996262e-15 -1.77635684e-15]
#
# numeric errors, sigma parameters
# [-8.88178420e-16 -2.22044605e-16]
#
# machine epsilon = 2.220446049250313e-16
#
# numeric epsilons, fit parameters:    [44. -8.]
#
# numeric epsilons, sigma parameters:  [-4. -1.]
